// happy-dom/cjs/PropertySymbol.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.referrer = exports.readyStateManager = exports.propagationStopped = exports.ownerElement = exports.ownerDocument = exports.mutationListeners = exports.observedAttributes = exports.observeMutations = exports.nextActiveElement = exports.namedItems = exports.listeners = exports.listenerOptions = exports.isValue = exports.isInPassiveEventListener = exports.isFirstWriteAfterOpen = exports.isFirstWrite = exports.indeterminate = exports.immediatePropagationStopped = exports.height = exports.internalId = exports.formNode = exports.exceptionObserver = exports.evaluateScript = exports.evaluateCSS = exports.entries = exports.end = exports.dirtyness = exports.destroy = exports.defaultView = exports.data = exports.currentTarget = exports.currentScript = exports.cssText = exports.contentType = exports.contentLength = exports.disconnectedFromDocument = exports.connectedToDocument = exports.disconnectedFromNode = exports.connectedToNode = exports.classList = exports.children = exports.childNodes = exports.checked = exports.callbacks = exports.cachedResponse = exports.buffer = exports.bodyBuffer = exports.asyncTaskManager = exports.activeElement = exports.abort = void 0;
exports.name = exports.clientTop = exports.clientLeft = exports.clientWidth = exports.clientHeight = exports.offsetTop = exports.offsetLeft = exports.offsetWidth = exports.offsetHeight = exports.isContentEditable = exports.contentEditable = exports.accessKeyLabel = exports.accessKey = exports.namespaceURI = exports.attributesProxy = exports.attributes = exports.scrollLeft = exports.scrollTop = exports.scrollWidth = exports.scrollHeight = exports.prefix = exports.tagName = exports.nodeType = exports.parentNode = exports.isConnected = exports.popup = exports.openerWindow = exports.openerFrame = exports.mutationObservers = exports.windowResizeListener = exports.window = exports.width = exports.value = exports.url = exports.updateSelectedness = exports.reportMutation = exports.unobserveMutations = exports.textAreaNode = exports.target = exports.style = exports.start = exports.shadowRoot = exports.setupVMContext = exports.selection = exports.selectedness = exports.selectNode = exports.rootNode = exports.resetSelection = exports.relList = exports.registry = void 0;
exports.sessionStorage = exports.screen = exports.navigator = exports.history = exports.location = exports.nodeStream = exports.registedClass = exports.localName = exports.setURL = exports.host = exports.mode = exports.content = exports.options = exports.played = exports.seekable = exports.seeking = exports.textTracks = exports.networkState = exports.ended = exports.error = exports.duration = exports.buffered = exports.preservesPitch = exports.defaultMuted = exports.muted = exports.defaultPlaybackRate = exports.playbackRate = exports.currentTime = exports.paused = exports.volume = exports.sheet = exports.files = exports.defaultChecked = exports.y = exports.x = exports.loading = exports.naturalWidth = exports.naturalHeight = exports.complete = exports.length = exports.elements = exports.returnValue = exports.validity = exports.validationMessage = exports.systemId = exports.publicId = exports.readyState = exports.implementation = exports.adoptedStyleSheets = exports.specified = void 0;
exports.assignedNodes = exports.slotAssignment = exports.serializable = exports.delegatesFocus = exports.clonable = exports.elementIdMap = exports.defaultValue = exports.areas = exports.top = exports.parent = exports.self = exports.selectedIndex = exports.attributeName = exports.getTokenList = exports.setNamedItem = exports.proxy = exports.namespaceItems = exports.getNamedItemKey = exports.getNamespaceItemKey = exports.dataset = exports.getFormControlNamedItem = exports.getFormControlItems = exports.computedStyle = exports.query = exports.affectsComputedStyleCache = exports.forms = exports.affectsCache = exports.cache = exports.elementArray = exports.nodeArray = exports.onRemoveAttribute = exports.onSetAttribute = exports.clearCache = exports.updateSheet = exports.styleNode = exports.selectedOptions = exports.items = exports.removeNamedItem = exports.clone = exports.settings = exports.capabilities = exports.constraints = exports.tracks = exports.replaceChild = exports.insertBefore = exports.removeChild = exports.appendChild = exports.cloneNode = exports.sandbox = exports.localStorage = void 0;
exports.windowInternalId = exports.browserFrames = exports.aborted = exports.destroyed = exports.globalObject = exports.detail = exports.type = exports.timeStamp = exports.eventPhase = exports.defaultPrevented = exports.cancelable = exports.bubbles = exports.composed = exports.popoverTargetElement = exports.canvas = exports.state = exports.illegalConstructor = exports.id = exports.language = exports.label = exports.kind = exports.activeCues = exports.cues = exports.srcObject = exports.sinkId = exports.remote = exports.mediaKeys = exports.controlsList = exports.track = exports.tBodies = exports.headers = exports.rows = exports.cells = exports.assignedToSlot = void 0;
exports.abort = Symbol('abort');
exports.activeElement = Symbol('activeElement');
exports.asyncTaskManager = Symbol('asyncTaskManager');
exports.bodyBuffer = Symbol('bodyBuffer');
exports.buffer = Symbol('buffer');
exports.cachedResponse = Symbol('cachedResponse');
exports.callbacks = Symbol('callbacks');
exports.checked = Symbol('checked');
exports.childNodes = Symbol('childNodes');
exports.children = Symbol('children');
exports.classList = Symbol('classList');
exports.connectedToNode = Symbol('connectedToNode');
exports.disconnectedFromNode = Symbol('disconnectedFromNode');
exports.connectedToDocument = Symbol('connectedToDocument');
exports.disconnectedFromDocument = Symbol('disconnectedFromDocument');
exports.contentLength = Symbol('contentLength');
exports.contentType = Symbol('contentType');
exports.cssText = Symbol('cssText');
exports.currentScript = Symbol('currentScript');
exports.currentTarget = Symbol('currentTarget');
exports.data = Symbol('data');
exports.defaultView = Symbol('defaultView');
exports.destroy = Symbol('destroy');
exports.dirtyness = Symbol('dirtyness');
exports.end = Symbol('end');
exports.entries = Symbol('entries');
exports.evaluateCSS = Symbol('evaluateCSS');
exports.evaluateScript = Symbol('evaluateScript');
exports.exceptionObserver = Symbol('exceptionObserver');
exports.formNode = Symbol('formNode');
exports.internalId = Symbol('internalId');
exports.height = Symbol('height');
exports.immediatePropagationStopped = Symbol('immediatePropagationStopped');
exports.indeterminate = Symbol('indeterminate');
exports.isFirstWrite = Symbol('isFirstWrite');
exports.isFirstWriteAfterOpen = Symbol('isFirstWriteAfterOpen');
exports.isInPassiveEventListener = Symbol('isInPassiveEventListener');
exports.isValue = Symbol('isValue');
exports.listenerOptions = Symbol('listenerOptions');
exports.listeners = Symbol('listeners');
exports.namedItems = Symbol('namedItems');
exports.nextActiveElement = Symbol('nextActiveElement');
exports.observeMutations = Symbol('observeMutations');
exports.observedAttributes = Symbol('observedAttributes');
exports.mutationListeners = Symbol('mutationListeners');
exports.ownerDocument = Symbol('ownerDocument');
exports.ownerElement = Symbol('ownerElement');
exports.propagationStopped = Symbol('propagationStopped');
exports.readyStateManager = Symbol('readyStateManager');
exports.referrer = Symbol('referrer');
exports.registry = Symbol('registry');
exports.relList = Symbol('relList');
exports.resetSelection = Symbol('resetSelection');
exports.rootNode = Symbol('rootNode');
exports.selectNode = Symbol('selectNode');
exports.selectedness = Symbol('selectedness');
exports.selection = Symbol('selection');
exports.setupVMContext = Symbol('setupVMContext');
exports.shadowRoot = Symbol('shadowRoot');
exports.start = Symbol('start');
exports.style = Symbol('style');
exports.target = Symbol('target');
exports.textAreaNode = Symbol('textAreaNode');
exports.unobserveMutations = Symbol('unobserveMutations');
exports.reportMutation = Symbol('reportMutation');
exports.updateSelectedness = Symbol('updateSelectedness');
exports.url = Symbol('url');
exports.value = Symbol('value');
exports.width = Symbol('width');
exports.window = Symbol('window');
exports.windowResizeListener = Symbol('windowResizeListener');
exports.mutationObservers = Symbol('mutationObservers');
exports.openerFrame = Symbol('openerFrame');
exports.openerWindow = Symbol('openerWindow');
exports.popup = Symbol('popup');
exports.isConnected = Symbol('isConnected');
exports.parentNode = Symbol('parentNode');
exports.nodeType = Symbol('nodeType');
exports.tagName = Symbol('tagName');
exports.prefix = Symbol('prefix');
exports.scrollHeight = Symbol('scrollHeight');
exports.scrollWidth = Symbol('scrollWidth');
exports.scrollTop = Symbol('scrollTop');
exports.scrollLeft = Symbol('scrollLeft');
exports.attributes = Symbol('attributes');
exports.attributesProxy = Symbol('attributesProxy');
exports.namespaceURI = Symbol('namespaceURI');
exports.accessKey = Symbol('accessKey');
exports.accessKeyLabel = Symbol('accessKeyLabel');
exports.contentEditable = Symbol('contentEditable');
exports.isContentEditable = Symbol('isContentEditable');
exports.offsetHeight = Symbol('offsetHeight');
exports.offsetWidth = Symbol('offsetWidth');
exports.offsetLeft = Symbol('offsetLeft');
exports.offsetTop = Symbol('offsetTop');
exports.clientHeight = Symbol('clientHeight');
exports.clientWidth = Symbol('clientWidth');
exports.clientLeft = Symbol('clientLeft');
exports.clientTop = Symbol('clientTop');
exports.name = Symbol('name');
exports.specified = Symbol('specified');
exports.adoptedStyleSheets = Symbol('adoptedStyleSheets');
exports.implementation = Symbol('implementation');
exports.readyState = Symbol('readyState');
exports.publicId = Symbol('publicId');
exports.systemId = Symbol('systemId');
exports.validationMessage = Symbol('validationMessage');
exports.validity = Symbol('validity');
exports.returnValue = Symbol('returnValue');
exports.elements = Symbol('elements');
exports.length = Symbol('length');
exports.complete = Symbol('complete');
exports.naturalHeight = Symbol('naturalHeight');
exports.naturalWidth = Symbol('naturalWidth');
exports.loading = Symbol('loading');
exports.x = Symbol('x');
exports.y = Symbol('y');
exports.defaultChecked = Symbol('defaultChecked');
exports.files = Symbol('files');
exports.sheet = Symbol('sheet');
exports.volume = Symbol('volume');
exports.paused = Symbol('paused');
exports.currentTime = Symbol('currentTime');
exports.playbackRate = Symbol('playbackRate');
exports.defaultPlaybackRate = Symbol('defaultPlaybackRate');
exports.muted = Symbol('muted');
exports.defaultMuted = Symbol('defaultMuted');
exports.preservesPitch = Symbol('preservesPitch');
exports.buffered = Symbol('buffered');
exports.duration = Symbol('duration');
exports.error = Symbol('error');
exports.ended = Symbol('ended');
exports.networkState = Symbol('networkState');
exports.textTracks = Symbol('textTracks');
exports.seeking = Symbol('seeking');
exports.seekable = Symbol('seekable');
exports.played = Symbol('played');
exports.options = Symbol('options');
exports.content = Symbol('content');
exports.mode = Symbol('mode');
exports.host = Symbol('host');
exports.setURL = Symbol('setURL');
exports.localName = Symbol('localName');
exports.registedClass = Symbol('registedClass');
exports.nodeStream = Symbol('nodeStream');
exports.location = Symbol('location');
exports.history = Symbol('history');
exports.navigator = Symbol('navigator');
exports.screen = Symbol('screen');
exports.sessionStorage = Symbol('sessionStorage');
exports.localStorage = Symbol('localStorage');
exports.sandbox = Symbol('sandbox');
exports.cloneNode = Symbol('cloneNode');
exports.appendChild = Symbol('appendChild');
exports.removeChild = Symbol('removeChild');
exports.insertBefore = Symbol('insertBefore');
exports.replaceChild = Symbol('replaceChild');
exports.tracks = Symbol('tracks');
exports.constraints = Symbol('constraints');
exports.capabilities = Symbol('capabilities');
exports.settings = Symbol('settings');
exports.clone = Symbol('clone');
exports.removeNamedItem = Symbol('removeNamedItem');
exports.items = Symbol('items');
exports.selectedOptions = Symbol('selectedOptions');
exports.styleNode = Symbol('styleNode');
exports.updateSheet = Symbol('updateSheet');
exports.clearCache = Symbol('clearCache');
exports.onSetAttribute = Symbol('onSetAttribute');
exports.onRemoveAttribute = Symbol('onRemoveAttribute');
exports.nodeArray = Symbol('nodeArray');
exports.elementArray = Symbol('elementArray');
exports.cache = Symbol('cache');
exports.affectsCache = Symbol('affectsCache');
exports.forms = Symbol('forms');
exports.affectsComputedStyleCache = Symbol('affectsComputedStyleCache');
exports.query = Symbol('query');
exports.computedStyle = Symbol('computedStyle');
exports.getFormControlItems = Symbol('getFormControlItems');
exports.getFormControlNamedItem = Symbol('getFormControlNamedItem');
exports.dataset = Symbol('dataset');
exports.getNamespaceItemKey = Symbol('getNamespaceItemKey');
exports.getNamedItemKey = Symbol('getNamedItemKey');
exports.namespaceItems = Symbol('namespaceItems');
exports.proxy = Symbol('proxy');
exports.setNamedItem = Symbol('setNamedItem');
exports.getTokenList = Symbol('getTokenList');
exports.attributeName = Symbol('attributeName');
exports.selectedIndex = Symbol('selectedIndex');
exports.self = Symbol('self');
exports.parent = Symbol('parent');
exports.top = Symbol('top');
exports.areas = Symbol('areas');
exports.defaultValue = Symbol('defaultValue');
exports.elementIdMap = Symbol('elementIdMap');
exports.clonable = Symbol('clonable');
exports.delegatesFocus = Symbol('delegatesFocus');
exports.serializable = Symbol('serializable');
exports.slotAssignment = Symbol('slotAssignment');
exports.assignedNodes = Symbol('assignedNodes');
exports.assignedToSlot = Symbol('assignedToSlot');
exports.cells = Symbol('cells');
exports.rows = Symbol('rows');
exports.headers = Symbol('headers');
exports.tBodies = Symbol('tBodies');
exports.track = Symbol('track');
exports.controlsList = Symbol('controlsList');
exports.mediaKeys = Symbol('mediaKeys');
exports.remote = Symbol('remote');
exports.sinkId = Symbol('sinkId');
exports.srcObject = Symbol('srcObject');
exports.cues = Symbol('cues');
exports.activeCues = Symbol('activeCues');
exports.kind = Symbol('kind');
exports.label = Symbol('label');
exports.language = Symbol('language');
exports.id = Symbol('id');
exports.illegalConstructor = Symbol('illegalConstructor');
exports.state = Symbol('state');
exports.canvas = Symbol('canvas');
exports.popoverTargetElement = Symbol('popoverTargetElement');
exports.composed = Symbol('composed');
exports.bubbles = Symbol('bubbles');
exports.cancelable = Symbol('cancelable');
exports.defaultPrevented = Symbol('defaultPrevented');
exports.eventPhase = Symbol('eventPhase');
exports.timeStamp = Symbol('timeStamp');
exports.type = Symbol('type');
exports.detail = Symbol('detail');
exports.globalObject = Symbol('globalObject');
exports.destroyed = Symbol('destroyed');
exports.aborted = Symbol('aborted');
exports.browserFrames = Symbol('browserFrames');
exports.windowInternalId = Symbol('windowInternalId');
